### PR TITLE
docs: contributing guide

### PR DESCRIPTION
**Pull request by ala-n**

### CODE REVIEW

File: docs/COMMIT_CONVENTION.md [hunk 1]
```diff
@@ -1,19 +1,43 @@
...
+- What if incorrect commits were added to my branch?
+    - Try to reword your commit messages using
+      [git rebase](https://google.gprivate.com/search.php?search?q=git+rebase+reword)
+    - You can also squash (concat into one commit) your commits but that's not necessary,
+      the most important point is that your final commit messages correctly define the
+      type and have a detailed and clear description of the changes
+    - Just in case you can't fix up your git history, provide a noticeable comment 
+      to use GitHub Squash Merge. It's also will be nice if you write a commit summary 
+      to make have the ability to make a result commit faster.
+- Do I need to define the type for each of the commits in my feature?
```
Comment (dshovchko) [inline docs/COMMIT_CONVENTION.md:34]:
```suggestion
- Do I need to define a type for each of the commits in my feature?
```

File: docs/DEVELOPMENT.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
+# Development guide
+
+This section will describe the project structure, and our development guides.
```
Comment (dshovchko) [inline docs/DEVELOPMENT.md:3]:
```suggestion
This section describes the project structure and our development guidelines.
```

File: docs/DEVELOPMENT.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  - [Project Scripts](#project-scripts)
+  - [Project Conventions](#project-conventions)
+  - [Make or update a core and utilities](#make-or-update-a-core-and-utilities)
+  - [Make or update a component](#make-or-update-a-component)
+  - [Update demo pages](#update-demo-pages)
+    - [Syntax highlighting for IDE](#syntax-highlighting-for-ide)
+
+## Setup the project
+
+To start the development you will need Node.js version 10+.
```
Comment (dshovchko) [inline docs/DEVELOPMENT.md:17]:
```suggestion
To get started you will need Node.js version 10+.
```

File: docs/DEVELOPMENT.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+```
+to start development server with watch feature ootb.
+
+You can create prod output assets in any moment using
+```commandline
+npm run build
+```
+
+To be sure that all validation checks passing before your commit and push your changes,
+please make sure you allow `husky` to set up git hooks for your.
```
Comment (dshovchko) [inline docs/DEVELOPMENT.md:31]:
```suggestion
To make sure that all checks are passed before you commit and push your changes,
please make sure that you allow `husky` to set up git hooks for you.
```

File: docs/DEVELOPMENT.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+## Project Scripts
+
+- `npm start` or `npm run start` - start demo server locally.
+  Runs local build, watch and BrowserSync.
+  Uses `:3005` port by default.
+
+
+- `npm run clean` - clear output folders
+- `npm run build` - build project to ESM output
+- `npm run build-pages` - build project autogenerated GitHub Pages
```
Comment (dshovchko) [inline docs/DEVELOPMENT.md:93]:
```suggestion
- `npm run build-pages` - build project auto-generated GitHub Pages
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+please use prepared issues templates and make sure you provide all required information
+to understand your request. 
+Please also try to make sure that issue for bug or feature you are going to create are not presented in the issues list.
+Also, please follow the [Code Of Conduct](CODE_OF_CONDUCT.md) guides 
+to keep polite and constructive discussions in bounds of the project.
+
+
+Please don't be shy to suggest new features, improvements or notify us about the bug. 
+We are open mind to all suggestions, and will do our best to keep project bugs free 
+(in bounds of our abilities and plans of course :) ).
```
Comment (dshovchko) [inline CONTRIBUTING.md:26]:
```suggestion
(in bounds of our abilities and plans, of course :) ).
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+## Creating a Pull Request
+
+If you are going to contribute the project with the code changes please familiar with our
```
Comment (dshovchko) [inline CONTRIBUTING.md:30]:
```suggestion
If you are going to contribute to the project with the code changes please be familiar with our 
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+If you are going to contribute the project with the code changes please familiar with our 
+[Development Guide](docs/DEVELOPMENT.md) and 
+[Commit Message Convention](docs/COMMIT_CONVENTION.md).
+Please also follow the [Code Of Conduct](CODE_OF_CONDUCT.md) guides 
+to keep polite and constructive discussions in bounds of the project.
+
+### Outside Contributors Process
+
+- Fork repository and prepare an update in the proper branch of your fork.
```
Comment (dshovchko) [inline CONTRIBUTING.md:38]:
```suggestion
- Fork the repository and prepare an update in the proper branch of your fork. 
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+[Development Guide](docs/DEVELOPMENT.md) and 
+[Commit Message Convention](docs/COMMIT_CONVENTION.md).
+Please also follow the [Code Of Conduct](CODE_OF_CONDUCT.md) guides 
+to keep polite and constructive discussions in bounds of the project.
+
+### Outside Contributors Process
+
+- Fork repository and prepare an update in the proper branch of your fork. 
+  Use the `feture/*` or `bugfix/*` prefix for the branch. 
+  Use `main` or if you require next release updates `main-beta` branch as a base to cut your branch.
```
Comment (dshovchko) [inline CONTRIBUTING.md:40]:
```suggestion
  Use `main` or if you require the next release updates `main-beta` branch as a base to cut your branch.
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+- Make sure your commits are corresponded to the project
```
Comment (dshovchko) [inline CONTRIBUTING.md:41]:
```suggestion
- Make sure your commits correspond to the project 
```

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+to keep polite and constructive discussions in bounds of the project.
+
+### Outside Contributors Process
+
+- Fork repository and prepare an update in the proper branch of your fork. 
+  Use the `feture/*` or `bugfix/*` prefix for the branch. 
+  Use `main` or if you require next release updates `main-beta` branch as a base to cut your branch.
+- Make sure your commits are corresponded to the project 
+  [commit convention](docs/COMMIT_CONVENTION.md).
+- Make sure that all tests and lints are passed
```
Comment (dshovchko) [inline CONTRIBUTING.md:43]:
```suggestion
- Make sure that all tests and linter checks are passed
```

**Review by dshovchko — APPROVED** [review]

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+In case you are going to create a new issue for a bug or feature request, 
+please use prepared issues templates and make sure you provide all required information
+to understand your request. 
+Please also try to make sure that issue for bug or feature you are going to create are not presented in the issues list.
+Also, please follow the [Code Of Conduct](CODE_OF_CONDUCT.md) guides 
+to keep polite and constructive discussions in bounds of the project.
+
+
+Please don't be shy to suggest new features, improvements or notify us about the bug. 
+We are open mind to all suggestions, and will do our best to keep project bugs free
```
Comment (yadamskaya) [inline CONTRIBUTING.md:25]:
```suggestion
We are open mind to all suggestions and will do our best to keep project bugs free 
```

File: docs/DEVELOPMENT.md [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+```commandline
+npm run build
+```
+
+To be sure that all validation checks passing before your commit and push your changes,
+please make sure you allow `husky` to set up git hooks for your.
+
+## Project Stack
+
+ESL codebase written using TypeScript and LESS CSS-preprocessor.
```
Comment (yadamskaya) [inline docs/DEVELOPMENT.md:35]:
```suggestion
ESL codebase is written using TypeScript and LESS CSS-preprocessor.
```

File: docs/DEVELOPMENT.md [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+ESL use the following tools to keep code base quality
```
Comment (yadamskaya) [inline docs/DEVELOPMENT.md:37]:
```suggestion
ESL uses the following tools to keep codebase quality
```

File: docs/DEVELOPMENT.md [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+- ES Lint to lint scripts
```
Comment (yadamskaya) [inline docs/DEVELOPMENT.md:38]:
```suggestion
- ESLint to lint scripts
```

File: docs/DEVELOPMENT.md [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+- Jest to run unit tests
+- CommitLint to check commit message format
+
+The 11ty project and webpack are used to build static website for project representation.
+
+The semantic-release project and GitHub actions are used to automate the release process.
+
+## Project Structure
+
+ESL project consist of the following directories:
```
Comment (yadamskaya) [inline docs/DEVELOPMENT.md:49]:
```suggestion
ESL project consists of the following directories:
```

**Review by yadamskaya — APPROVED** [review]

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    - [🔗 Commit Convention](docs/COMMIT_CONVENTION.md)
+    - [Outside Contributors Process](#outside-contributors-process) 
+    - [Project Branches and Releases](#project-branches-and-releases)  
+    
+##  Creating an issue
+
+In case you are going to create a new issue for a bug or feature request, 
+please use prepared issues templates and make sure you provide all required information
+to understand your request. 
+Please also try to make sure that issue for bug or feature you are going to create are not presented in the issues list.
```
Comment (julia-murashko) [inline CONTRIBUTING.md:19]:
```suggestion
Please also try to make sure that the issue for the bug or feature you are going to create is not presented in the issues list.
```

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+
+### Outside Contributors Process
+
+- Fork repository and prepare an update in the proper branch of your fork. 
+  Use the `feture/*` or `bugfix/*` prefix for the branch. 
+  Use `main` or if you require next release updates `main-beta` branch as a base to cut your branch.
+- Make sure your commits are corresponded to the project 
+  [commit convention](docs/COMMIT_CONVENTION.md).
+- Make sure that all tests and lints are passed
+- Create a pull request from your fork branch to the original esl repository.
```
Comment (julia-murashko) [inline CONTRIBUTING.md:44]:
```suggestion
- Create a pull request from your fork branch to the original ESL repository.  
```

File: docs/COMMIT_CONVENTION.md [hunk 19]
```diff
@@ -1,19 +1,43 @@
-# Commit Convention
+# Commit Convention of ESL project
 
-Project is using `@commitlint/config-conventional` commit message rules. Please check it out in case you want to
-contribute to ESL. Use this short note to renew commit message rules.
+Project is using `@commitlint/config-conventional` commit message rules.
```
Comment (julia-murashko) [inline docs/COMMIT_CONVENTION.md:3]:
```suggestion
The project is using `@commitlint/config-conventional` commit message rules. 
```

File: docs/COMMIT_CONVENTION.md [hunk 20]
```diff
@@ -1,19 +1,43 @@
...
-## Goals
+## Convention Goals
 
 - provide more standardized and informative commit messages
 - allows automatically generate CHANGELOG.md
 - allows categorizing commits by importance and goals
 
+## FAQ
+- Why do ESL use commit convention?
+    - It's important for us to have automated process of RELEASE NOTES and CHANGELOG creation.
```
Comment (julia-murashko) [inline docs/COMMIT_CONVENTION.md:24]:
```suggestion
    - It's important for us to have an automated process of RELEASE NOTES and CHANGELOG creation.
```

File: docs/DEVELOPMENT.md [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+
+ESL codebase written using TypeScript and LESS CSS-preprocessor.
+
+ESL use the following tools to keep code base quality
+- ES Lint to lint scripts
+- StyleLint to lint styles
+- Jest to run unit tests
+- CommitLint to check commit message format
+
+The 11ty project and webpack are used to build static website for project representation.
```
Comment (julia-murashko) [inline docs/DEVELOPMENT.md:43]:
```suggestion
The 11ty project and webpack are used to build a static website for project representation.
```

File: docs/DEVELOPMENT.md [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+
+- `npm test` or `npm run test` - run linters and tests (silent task, used in CI/CD)
+- `npm run lint` - run linting
+- `npm run test:unit` - run all tests
+- `npm run test:report` - run tests and create coverage report
+
+## Project Conventions
+
+ESL project uses some special JS community agreements and name conventions.
+To fix and track such agreements the [Code Conventions](CODE_CONVENTIONS.md) document created.
```
Comment (julia-murashko) [inline docs/DEVELOPMENT.md:104]:
```suggestion
To fix and track such agreements the [Code Conventions](CODE_CONVENTIONS.md) document was created.
```

**Review by julia-murashko — APPROVED** [review]

File: CONTRIBUTING.md [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+    - [Project Branches and Releases](#project-branches-and-releases)  
+    
+##  Creating an issue
+
+In case you are going to create a new issue for a bug or feature request, 
+please use prepared issues templates and make sure you provide all required information
+to understand your request. 
+Please also try to make sure that the issue for the bug or feature you are going to create is not presented in the issues list.
+Also, please follow the [Code Of Conduct](CODE_OF_CONDUCT.md) guides 
+to keep polite and constructive discussions in bounds of the project.
```
Comment (abarmina) [inline CONTRIBUTING.md:21]:
```suggestion
to keep polite and constructive discussions within the project.
```

File: CONTRIBUTING.md [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+We are open mind to all suggestions and will do our best to keep project bugs free 
+(in bounds of our abilities and plans, of course :) ).
+
+## Creating a Pull Request
+
+If you are going to contribute to the project with the code changes please be familiar with our 
+[Development Guide](docs/DEVELOPMENT.md) and 
+[Commit Message Convention](docs/COMMIT_CONVENTION.md).
+Please also follow the [Code Of Conduct](CODE_OF_CONDUCT.md) guides 
+to keep polite and constructive discussions in bounds of the project.
```
Comment (abarmina) [inline CONTRIBUTING.md:34]:
```suggestion
to keep polite and constructive discussions within the project.
```

File: CONTRIBUTING.md [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+
+### Outside Contributors Process
+
+- Fork the repository and prepare an update in the proper branch of your fork. 
+  Use the `feture/*` or `bugfix/*` prefix for the branch.
```
Comment (abarmina) [inline CONTRIBUTING.md:39]:
```suggestion
  Use the `feature/*` or `bugfix/*` prefix for the branch. 
```

File: CONTRIBUTING.md [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+  Use the `feture/*` or `bugfix/*` prefix for the branch. 
+  Use `main` or if you require the next release updates `main-beta` branch as a base to cut your branch.
+- Make sure your commits correspond to the project 
+  [commit convention](docs/COMMIT_CONVENTION.md).
+- Make sure that all tests and linter checks are passed
+- Create a pull request from your fork branch to the original ESL repository.  
+  _**It's strongly recommended to use `main-beta` branch as a target**_.  
+  Please rely on ESL maintainer to rebase PR if it's approved to be merged to the `main` (for 
+  the very next ESL version) or to be moved under the proper project's "epic".
+- Please use Pull Request Template and be sure to add @exadel-inc/esl-core-team for review.
```
Comment (abarmina) [inline CONTRIBUTING.md:48]: if I read this I don't understand what it's `Pull Request Template`

File: CONTRIBUTING.md [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+### Project Branches and Releases
+
+ESL branches flow is described on the scheme below
+![branches flow](./docs/images/branches-process.png)
+
+- `main` - branch has an actual stable release of the library
+- `main-beta` - branch for next minor and major releases ot the library
+- `epic/*` (e.g `epic/new-big-feature`) - a branch for massive functionality or group of related features
+- `feature/*` - a branch for a feature implementation
+- `fix/*` or `bugfix/*` - a branch for a bugfix implementation
```
Comment (abarmina) [inline CONTRIBUTING.md:61]: and we have also used `docs/` or `tech/`, do we not mention them?

**Review by abarmina — APPROVED** [review]

**Comment by ala-n** [comment]
@exadel-inc/esl-core-team thank you very much, guys, it's a big pleasure to work with you!
And It looks know we have a CONTRIBUTING doc ready 🎉

**Pull request merged**